serokell-style.md

```diff
@@ -288,372 +300,261 @@ Prefer two-line imports for such standard containers.
-arithmetics and everything is `Integer` then type signatures can look cumbersome.
+signatures help avoid cryptic type errors. You _may_ choose not to specify them in
+cases like working with pure arithmetics where everything is `Integer`, and explicit
+type signatures look cumbersome.
 
 > You most likely need `-XExplicitForAll` and `-XScopedTypeVariables` extensions
 > to write polymorphic types of functions inside `where`.
 
-Specialize function type signature for concrete types if you're using this function
-with only one type for each argument. Otherwize you should use more polymorphic
-version. Compiler can optimize specialized functions better
-(TODO: link to haskell manual) and meaning of this function may be clearer.
-Use this rule unless you are the library creator and want your library to be abstract
-as possible.
+You _should_ avoid overly general signatures for functions that are actually used
+with only one type for each parameter. If you need the polymorhic version (i.e.
+if you are instantiating it more than once or if you are writing a library), you
+_may_ use GHC's
+[`SPECIALIZE` pragma](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#specialize-pragma).
 
-It is allowed to omit parentheses for only one type class constraint.
+You _should_ omit parentheses if you have only one constraint.
 
-If function type signature is very long then place type of each argument under
-its own line with respect to alignment.
+If function type signature is very long, you _should_ place the type of each argument
+on its own line, and also align constraints similarly. Example:
 
 ```haskell
 putValueInState
-    :: MonadIO m
-    => UserState
-    -> Maybe Int
-    -> AppConfig
-    -> (Int -> m ())
-    -> m ()
+  :: (MonadIO m, WithLogger m)
+  => UserState
+  -> Maybe Int
+  -> AppConfig
+  -> (Int -> m ())
+  -> m ()
+```
+
+If there are a lot of constraints, or they are very long, you _may_ put them
+in a separate `type` definition, or format them as follows:
+
+```haskell
+parseTree
+  :: ( KnownSpine components
+     , AllConstrained (ComponentTokenizer components) components
+     , AllConstrained (ComponentTokenToLit components) components
+     )
+  => Text
+  -> Either (ParseError components) (Expr ParseTreeExt CommandId components)
 ```
 
-If the line with argument names is too big then put each argument on its own line
+If the line with argument names is too long, you _should_ put each argument
+on a separate line with the usual 2-space indentation.
 and separate it somehow from body section.
 
 ```haskell
 putValueInState
-    userState
-    mValue@(Just x)
-    Config{..}        -- { should go after ctor name without space
-    valueModificator
-  = do
-    <code goes here>
+  userState
+  mValue@(Just x)
+  Config{..}        -- { should go after ctor name without space
+  valueModificator
+ = do               -- note how this line uses 1-space indentation
```

Comment:
> allowing 4-space for params and 2-space for `= do` here?

That's what I'd do, but @Pastafarianist is concerned that it means adding parameters to a function might lead to reindenting the entire function body because of this rule.

Reply:
Assume we have

```
loadAndProcessSomeData param1 param2 param3 = do
  stmt1
  stmt2
  stmt3
```

We add more parameters so that they no longer fit on one line:

```
loadAndProcessSomeData
    param1
    param2
    param3
    param4
    param5
  = do
    stmt1
    stmt2
    stmt3
```

We want to avoid reindentation of statements.

One way is to allow 1-space indent of `= do`, another way is to avoid hanging `do` altogether. We could advocate this style instead:

```
loadAndProcessSomeData param1 param2 param3 =
  do
    stmt1
    stmt2
    stmt3

loadAndProcessSomeData
    param1
    param2
    param3
    param4
    param5
  =
  do
    stmt1
    stmt2
    stmt3
```

Then the `do` block is not touched at all.

Reply:
Seems like 1-space indentation is a simpler solution as it is closer to our current style.

Reply:
If we're using 2 spaces, let's use 2 spaces everywhere, not this half-size 1-space indentation to fix one particular problem (there are probably other cases that require re-indentation and aren't addressed).

Reply:
What I mean is that we never use this hanging `do` you are proposing. 1-space indentation for this particular case is mimicking a workaround we've been implementing with the 4-space indentation our current style guide enforces. I think a major change (which is itself debatable) is worse than a small one with a workaround.

Reply:
> 1-space indentation for this particular case is mimicking a workaround we've been implementing with the 4-space indentation our current style guide enforces.

Sure, that's something I pointed out, it still doesn't mean we should keep doing this. We don't have to do the larger change of avoiding hanging `do`. I just checked and both of these work:

```
fn
  param1
  param2
  = do
  stmt1
  stmt2

fn
  param1
  param2
    = do
  stmt1
  stmt2
```

No need for 1-space indent.

Reply:
Awaiting results of vote in Slack.